src/libopensc/card.c

```diff
@@ -507,8 +507,14 @@ int sc_lock(sc_card_t *card)
 	}
 
 	/* give card driver a chance to do something when reader lock first obtained */
-	if (r == 0 && reader_lock_obtained == 1  && card->ops->card_reader_lock_obtained)
+	if (r == 0 && reader_lock_obtained == 1  && card->ops->card_reader_lock_obtained) {
 		r = card->ops->card_reader_lock_obtained(card, was_reset);
+		if (r != 0) {
+			if (card->reader->ops->unlock != NULL)
+				r = card->reader->ops->unlock(card->reader);
+			card->lock_count--;
```

Comment:
Have we seen any problems in this code? Is this just a fuzz issue?

These drivers set a version of card_reader_lock_obtained:
```
card-asepcos.c:	asepcos_ops.card_reader_lock_obtained = asepcos_card_reader_lock_obtained;
card-authentic.c:	authentic_ops.card_reader_lock_obtained = authentic_card_reader_lock_obtained;
card-coolkey.c:	coolkey_ops.card_reader_lock_obtained = coolkey_card_reader_lock_obtained;
card-gemsafeV1.c:	gemsafe_ops.card_reader_lock_obtained = gemsafe_card_reader_lock_obtained;
card-gids.c:	gids_ops.card_reader_lock_obtained = gids_card_reader_lock_obtained;
card-isoApplet.c:	isoApplet_ops.card_reader_lock_obtained = isoApplet_card_reader_lock_obtained;
card-jpki.c:	jpki_ops.card_reader_lock_obtained = jpki_card_reader_lock_obtained;
card-muscle.c:	muscle_ops.card_reader_lock_obtained = muscle_card_reader_lock_obtained;
card-nqApplet.c:	nqapplet_operations.card_reader_lock_obtained;
card-openpgp.c:	pgp_ops.card_reader_lock_obtained = pgp_card_reader_lock_obtained;
card-piv.c:	piv_ops.card_reader_lock_obtained = piv_card_reader_lock_obtained;
```
So before doing anything, these should be examined to see how they handle an error before returning and do any of them calls sc_unlock. Do any expect a failure in their` reader_lock_obtained` routine to be ignored?

I expect an error returned from  `reader_lock_obtained` would be considered fatal and the error would then be returned to the caller of `sc_lock`.  Do all calls to `sc_lock` check the return code?

IMHO any error return by `reader_lock_obtained` should be considered fatal, i.e. it could not be handled in `reader_lock_obtained` and needs to be passed back up the stack.
 
The call to sc_lock could come from anywhere, not just the card driver. And any error from  card_reader_lock_obtained
is returned as the return code of the sc_lock.  So do all calls to sc_lock check the return code?

I think we should consider calling `card->ops->un_lock` if `reader_lock_obtained` fails but both 
 `r = card->reader->ops->unlock(card->reader);` and `card->lock_count--;` should not be done, as `unlock` will decrement the `card->lock_count` so we don't want to do it twice.

`card.c` and `card-gemsafeV1.c` are the only places `card->lock_count` is decremented.

Reply:
OK, looking at the context, I think I know where this PR is headed.

In the original code, the reader is locked (`card->reader->ops->unlock`) and `r`carries the result until it is returned to the caller. Then `r` is overwritten by `card->reader->ops->unlock`, which indicates to the caller that locking failed. Hence, the lock count should not have been incremented. So this PR is good in the sense that it fixes the lock count to match the return value of the function.

However, I still would still consider `card->ops->card_reader_lock_obtained` to non-fatal. Anyway, for the sake of moving forward, let's keep this PR as is.

Reply:
Yes, thank you for looking into it. The fuzz issue was related to lock count - when `sc_lock()` fails and its return value is checked, the `sc_disconnect_card()` function leaks the memory, since `card->lock_count` remains non-zero (should have mentioned it in commit message, sorry).

Reply:
Ok, mentioned in the code and commit message.

Reply:
Thanks. Looks good to me.